File: src/reverse-word-in-a-sentence.cpp

// Program to reverse the word in a sentence... String sentence is given by the user.

/*
 we are first reversing each word and the nreversing the whole string.
 Example : This is the reverse senetence. (Original String)
           sihT si eht esrever .ecnetenes (After reversing each word)
           sentence. reverse the is This  (After reversing the whole string.)
*/

#include<stdio.h>

/* function to reverse a string */
void reverse(char *start, char *end);

/*Function to reverse words*/
void reversewords(char *s)
{
  char *word = s;
  char *temp = s;

  /*start reversing the individual words*/
  while( *temp )
  {
    temp++;
    if (*temp == '\0')
    {
      reverse(word, temp-1);
    }
    else if(*temp == ' ')
    {
      reverse(word, temp-1);
      word = temp+1;
    }
  }
  reverse(s, temp-1);/*Reverse the resultant string from beginning to end including spaces*/
}

/*Function to reverse a string this is similar to earlier blog post to reverse a string */
void reverse(char *start, char *end)
{
  char temp;
  while (start < end)
  {
    temp = *start;
    *start++ = *end;
    *end-- = temp;
  }
}

int main()
{
  char s[] = "This is the reverse sentence.";
  char *temp = s;
  printf("Original string: %s\n",s);
  reversewords(s);
  printf("Reverse word string: %s\n", s);
  return 0;
}

File: src/reverse-word-in-a-sentence.java
// Program to reverse the word in a sentence... String sentence is given by the user.

/*
 we are first reversing each word and the nreversing the whole string.
 Example : This is the reverse senetence. (Original String)
           sihT si eht esrever .ecnetenes (After reversing each word)
           sentence. reverse the is This  (After reversing the whole string.)
*/

public class reveerseWord
{
  public staatic void main(String args[])
  {
    String str = "This is the reverse sentence";
    s = doIt(s);
    System.out.println("Reverse sentence : " + s);
  }

  void doIt(String str)
  {
    for(int i=0, s=0, e=0; i<length; i++)
    {
      if(str.charAt(i)==' ')
      {
        e = i-1;
        reerse(s,e);
        s = i+1;
      }
      if(str.charAt(i)=='\0')
      {
        e = i-1;
        reverse(s,e);
      }
    }

    reverse(0,length-1);
  }

  void reverse(int start, int end)
  {
    char temp;

    while(start < end)
    {
      temp = s[start];
      s[start] = s[end];
      s[end] = temp;
      start++;
      end--;
    }
  }

}
